File: src/main/java/com/tms/csp/fm/dnnf/DOr.java
package com.tms.csp.fm.dnnf;


import com.google.common.collect.ImmutableSet;
import com.tms.csp.argBuilder.ArgBuilder;
import com.tms.csp.ast.*;
import com.tms.csp.fm.dnnf.models.Solution;
import com.tms.csp.fm.dnnf.models.Solutions;
import com.tms.csp.fm.dnnf.products.Cube;
import com.tms.csp.util.varSets.VarSet;

import java.math.BigInteger;
import java.util.Set;

import static com.google.common.base.Preconditions.checkNotNull;

public class DOr extends Or {

    private static final String O_SP = "O ";

    private Exp smooth;

    private Boolean isSmooth;
    private Integer value;  //for counting graph

    public DOr(Space space, int id, Exp[] args) {
        super(space, id, args);
        for (Exp arg : args) {
            assert arg.isDnnf() : arg;
        }
    }

    public Exp getSmooth() {
        if (smooth == null) {
            if (isSmooth()) {
                smooth = this;
            } else {
                smooth = computeSmooth();
            }
        }
        return smooth;
    }

    private Exp computeSmooth() {
        assert smooth == null;
        if (isSmooth != null && isSmooth) {
            smooth = this;
            return this;
        }

        ArgBuilder b = new ArgBuilder(_space, op());

        VarSet parentCareVars = getVars();

        for (Exp child : args) {


            child = child.getSmooth();


            VarSet childCareVars = child.getVars();
            VarSet dontCares = parentCareVars.minus(childCareVars);


            Exp smoothChild;
            if (dontCares.isEmpty()) {
                smoothChild = child;
            } else {
                smoothChild = child.smooth(dontCares);
            }


            b.addExp(smoothChild);
        }


        Exp smooth = b.mk();
        smooth._setIsSmooth();

        return smooth;

    }

    @Override
    public Exp project(VarSet outVars) {
        checkNotNull(outVars);

        VarSet vars = this.getVars();
        if (outVars.containsAllVars(vars)) {
            return this;
        }

        ArgBuilder b = new ArgBuilder(_space, Op.DOr);


        for (int i = 0; i < args.length; i++) {
            Exp arg = args[i];
            Exp s = arg.project(outVars);

            if (s.isTrue()) {
                return mkTrue();
            } else if (s.isFalse()) {
                //skip
            } else {
                assert s.isOpen();
                b.addExp(s);

            }

        }

        return b.mk();
    }




    public boolean isOr() {
        return true;
    }


    @Override
    public void serializeTinyDnnf(Ser a) {
        a.append('O');
        a.append(' ');
        serializeArgsTinyDnnf(a);
    }


    @Override
    public boolean isDOr() {
        return true;
    }

    @Override
    public Exp toDnnf() {
        return this;
    }

    @Override
    public Op getOp() {
        return Op.DOr;
    }

    @Override
    public DOr asDOr() {
        return this;
    }

    @Override
    public boolean checkDnnf() {
        checkDeterministic();
        for (Exp arg : args) {
            arg.checkDnnf();
        }
        return true;
    }

    public boolean checkDeterministic() {
        return true;
    }

    public Exp flatten() {
        if (isFlat()) {
            return this;
        }


        ArgBuilder b = new ArgBuilder(_space, Op.DOr);
        for (Exp arg : args) {
            arg = arg.flatten();
            if (arg.isOr()) {
                for (Exp aa : arg.getArgs()) {
                    aa = aa.flatten();
                    b.addExp(aa);
                }
            } else {
                b.addExp(arg);
            }
        }


//        for (Exp arg : retVal) {
//            assert !arg.isNested(this);
//        }

        return b.mk();
    }


    public Set<Cube> computeCubesRough() {
        VarSet parentCareVars = getVars();

        ImmutableSet.Builder<Cube> b = ImmutableSet.builder();
        for (Exp arg : args) {
            VarSet argVars = arg.getVars();
            VarSet dcVars = parentCareVars.minus(argVars);

            if (arg.isLit()) {
                Solution solution = new Solution(_space, arg.asLit().asCube(), dcVars);
                b.addAll(solution);
            } else {
                Set<Cube> argCubes = arg.getCubesRough();
                Set<Cube> cubes = new Solutions(_space, argCubes, dcVars);
                b.addAll(cubes);
            }

        }
        return b.build();
    }


    public Set<Cube> computeCubesSmooth() {
        ImmutableSet.Builder<Cube> b = ImmutableSet.builder();
        for (Exp arg : args) {
            if (arg.isLit()) {
                b.add(arg.asLit().asCube());
            } else {
                Set<Cube> argCubes = arg.getCubesSmooth();
                b.addAll(argCubes);
            }
        }
        return b.build();
    }

    public BigInteger computeSatCount() {
        BigInteger satCount = BigInteger.ZERO;
        for (Exp arg : args) {
//            satCount += arg.getSatCount();
            satCount = satCount.add(arg.getSatCount());
        }
        return satCount;
    }


    @Override
    public Set<Cube> computeCubesNoVarSet() {
        VarSet parentCareVars = getVars();

        ImmutableSet.Builder<Cube> b = ImmutableSet.builder();
        for (Exp arg : args) {
            VarSet argVars = arg.getVars();
            VarSet dcVars = parentCareVars.minus(argVars);

            if (arg.isLit()) {
                Solution solution = new Solution(_space, arg.asLit().asCube(), dcVars);
                b.addAll(solution);
            } else {
                Set<Cube> argCubes = arg.getCubesSmooth();
                Set<Cube> cubes = new Solutions(_space, argCubes, dcVars);
                b.addAll(cubes);
            }

        }
        //        return new DisjointCubeSet(b);
        return b.build();
    }

    @Override
    public int getCubeCount() {
        return computeCubeCount();
    }


    public int computeCubeCount() {
        VarSet parentCareVars = getVars();

        ImmutableSet.Builder<Cube> b = ImmutableSet.builder();
        for (Exp arg : args) {
            VarSet argVars = arg.getVars();
            VarSet dcVars = parentCareVars.minus(argVars);

            if (arg.isLit()) {
                Solution solution = new Solution(_space, arg.asLit().asCube(), dcVars);
                b.addAll(solution);
            } else {
                Set<Cube> argCubes = arg.getCubesSmooth();
                Set<Cube> cubes = new Solutions(_space, argCubes, dcVars);
                b.addAll(cubes);
            }

        }
        //        return new DisjointCubeSet(b);
        return b.build().size();
    }


    /**
     * replace unmatched lits with: lit or (!lit and false)
     */
    public Exp litMatch() {
        Space space = getSpace();

        ImmutableSet.Builder<Exp> newArgs = ImmutableSet.builder();

        for (Exp arg : args) {
            Exp newArg = arg.litMatch();
            newArgs.add(newArg);
        }

        Exp retVal = space.mkDOr(newArgs.build());


        return retVal;

    }

    @Override
    public void _setIsSmooth() {
        //        assert smooth == null;
//        assert isSmooth == null;
        smooth = this;
        isSmooth = true;
    }

    @Override
    public boolean isSmooth() {
        if (isSmooth == null) {
            isSmooth = computeIsSmooth();
        }
        return isSmooth;
    }

    private boolean computeIsSmooth() {
        VarSet parentVars = getVars();
        for (Exp child : args) {
            VarSet childVars = child.getVars();
            if (!parentVars.equals(childVars)) {
                return false;
            }
            if (!child.isSmooth()) {
                return false;
            }
        }
        return true;
    }


    @Override
    public int computeValue(Cube cube) {
        int totalValue = 0;
        for (Exp child : args) {
            int childValue = child.computeValue(cube);
            totalValue += childValue;
        }
        return totalValue;
    }

    @Override
    public int computeValue() {
        assert _space.pics != null;
        if (this.value == null) {
            int totalValue = 0;
            for (Exp child : args) {
                int childValue = child.computeValue();
                totalValue += childValue;
            }
            this.value = totalValue;
        }
        return this.value;
    }

    public boolean computeSat(Lit lit) {
        for (Exp arg : args) {
            boolean sat = arg.computeSat(lit);
            if (sat) return true;
        }
        return false;
    }

    public boolean computeSat(Cube cube) {
        for (Exp arg : args) {
            boolean sat = arg.computeSat(cube);
            if (sat) return true;
        }
        return false;
    }

    public boolean computeSat(VarSet trueVars) {
        for (Exp arg : args) {
            boolean sat = arg.computeSat(trueVars);
            if (sat) return true;
        }
        return false;
    }

    public long computeSatCount1(Lit lit) {
        long satCount = 0;

        VarSet parentCareVars = getVars();

        for (Exp child : args) {

            long satCountWithOutDcs = child.computeSatCount1(lit);

            VarSet childCareVars = child.getVars();
            VarSet dontCares = parentCareVars.minus(childCareVars);

            int dcCount = dontCares.size();

            long dcSatCount = Exp.computeDcPermCount(dcCount);

            if (dcSatCount < 0) {
                throw new IllegalStateException();
            }

            long satCountWithDcs = satCountWithOutDcs * dcSatCount;

            satCount += satCountWithDcs;


        }

        return satCount;
    }

    @Override
    public boolean computeIsSat() {
        for (Exp arg : args) {
            boolean sat = arg.isSat();
            if (sat) {
                return true;
            }
        }
        return false;
    }


    @Override
    public Exp copyToOtherSpace(Space destSpace) {
        if (getSpace() == destSpace) {
            return this;
        }
        return copyArgsExpToOtherSpace(destSpace, op(), argIt());
    }

    @Override
    public boolean isDnnf() {
        return true;
    }

//    @Override
//    public Set<Lit> getLits() {
//        ImmutableSet.Builder<Lit> b = ImmutableSet.builder();
//        for (Exp arg : args) {
//            b.addAll(arg.getLits());
//        }
//        return b.build();
//    }

//    @Override
//    public boolean isLitMatched() {
//        for (Exp arg : args) {
//            if (!arg.isLitMatched()) {
//                return false;
//            }
//        }
//
//        return true;
//    }

}

File: src/main/java/com/tms/csp/fm/dnnf/DOr.kt
package com.tms.csp.fm.dnnf


import com.google.common.base.Preconditions.checkNotNull
import com.google.common.collect.ImmutableSet
import com.tms.csp.argBuilder.ArgBuilder
import com.tms.csp.ast.*
import com.tms.csp.fm.dnnf.models.Solution
import com.tms.csp.fm.dnnf.models.Solutions
import com.tms.csp.fm.dnnf.products.Cube
import com.tms.csp.util.varSets.VarSet
import java.math.BigInteger

class DOr(space: Space, id: Int, args: Array<Exp>) : Or(space, id, args) {

    private var _smooth: Exp? = null

    private var _isSmooth: Boolean? = null
    private var _value: Int? = null  //for counting graph


    override val isOr: Boolean get() = true


    override val isDOr: Boolean get() = true

    override val op: Op get() = Op.DOr

    override val cubeCount: Int get() = computeCubeCount()

    override val isDnnf: Boolean
        get() = true

    init {
        for (arg in args) {
            assert(arg.isDnnf) { arg }
        }
    }

    override val smooth: Exp
        get() {
            if (_smooth == null) {
                if (isSmooth) {
                    _smooth = this
                } else {
                    _smooth = computeSmooth()
                }
            }
            return _smooth!!
        }

    private fun computeSmooth(): Exp {
        assert(_smooth == null)
        if (_isSmooth != null && _isSmooth!!) {
            _smooth = this
            return this
        }

        val b = ArgBuilder(space, op())

        val parentVars = vars

        for (arg in args) {

            val child = arg.smooth

            val childVars = child.vars
            val dontCares = parentVars.minus(childVars)

            val smoothChild: Exp = if (dontCares.isEmpty()) {
                child
            } else {
                child.smooth(dontCares)
            }

            b.addExp(smoothChild)
        }


        val smooth = b.mk()
        smooth._setIsSmooth()

        return smooth

    }

    override fun project(outVars: VarSet): Exp {
        checkNotNull(outVars)

        val vars = this.vars
        if (outVars.containsAllVars(vars)) {
            return this
        }

        val b = ArgBuilder(_space, Op.DOr)


        for (i in 0 until _args.size) {
            val arg = _args[i]
            val s = arg.project(outVars)

            if (s.isTrue) {
                return mkTrue()
            } else if (s.isFalse) {
                //skip
            } else {
                assert(s.isOpen)
                b.addExp(s)

            }

        }

        return b.mk()
    }


    override fun serializeTinyDnnf(a: Ser) {
        a.append('O')
        a.append(' ')
        serializeArgsTinyDnnf(a)
    }

    override fun toDnnf(): Exp {
        return this
    }

    override fun asDOr(): DOr {
        return this
    }

    override fun checkDnnf(): Boolean {
        checkDeterministic()
        for (arg in args) {
            arg.checkDnnf()
        }
        return true
    }

    fun checkDeterministic(): Boolean {
        return true
    }

    override fun flatten(): Exp {
        if (isFlat) {
            return this
        }


        val b = ArgBuilder(_space, Op.DOr)
        for (arg in args) {
            val argf = arg.flatten()
            if (argf.isOr) {
                for (aa in argf.args) {
                    val aaf = aa.flatten()
                    b.addExp(aaf)
                }
            } else {
                b.addExp(arg)
            }
        }


        //        for (Exp arg : retVal) {
        //            assert !arg.isNested(this);
        //        }

        return b.mk()
    }


    override fun computeCubesRough(): Set<Cube> {
        val parentCareVars = vars

        val b = ImmutableSet.builder<Cube>()
        for (arg in args) {
            val argVars = arg.vars
            val dcVars = parentCareVars.minus(argVars)

            if (arg.isLit) {
                val solution = Solution(_space, arg.asLit.asCube, dcVars)
                b.addAll(solution)
            } else {
                val argCubes = arg.cubesRough
                val cubes = Solutions(_space, argCubes, dcVars)
                b.addAll(cubes)
            }

        }
        return b.build()
    }


    override fun computeCubesSmooth(): Set<Cube> {
        val b = ImmutableSet.builder<Cube>()
        for (arg in args) {
            if (arg.isLit) {
                b.add(arg.asLit.asCube)
            } else {
                val argCubes = arg.cubesSmooth
                b.addAll(argCubes)
            }
        }
        return b.build()
    }

    override fun computeSatCount(): BigInteger {
        var satCount = BigInteger.ZERO
        for (arg in args) {
            //            satCount += arg.getSatCount();
            satCount = satCount.add(arg.satCount)
        }
        return satCount
    }


    override fun computeCubesNoVarSet(): Set<Cube> {
        val parentCareVars = vars

        val b = ImmutableSet.builder<Cube>()
        for (arg in args) {
            val argVars = arg.vars
            val dcVars = parentCareVars.minus(argVars)

            if (arg.isLit) {
                val solution = Solution(_space, arg.asLit.asCube, dcVars)
                b.addAll(solution)
            } else {
                val argCubes = arg.cubesSmooth
                val cubes = Solutions(_space, argCubes, dcVars)
                b.addAll(cubes)
            }

        }
        //        return new DisjointCubeSet(b);
        return b.build()
    }


    fun computeCubeCount(): Int {
        val parentCareVars = vars

        val b = ImmutableSet.builder<Cube>()
        for (arg in args) {
            val argVars = arg.vars
            val dcVars = parentCareVars.minus(argVars)

            if (arg.isLit) {
                val solution = Solution(_space, arg.asLit.asCube, dcVars)
                b.addAll(solution)
            } else {
                val argCubes = arg.cubesSmooth
                val cubes = Solutions(_space, argCubes, dcVars)
                b.addAll(cubes)
            }

        }
        //        return new DisjointCubeSet(b);
        return b.build().size
    }


    /**
     * replace unmatched lits with: lit or (!lit and false)
     */
    override fun litMatch(): Exp {
        val space = space

        val newArgs = ImmutableSet.builder<Exp>()

        for (arg in args) {
            val newArg = arg.litMatch()
            newArgs.add(newArg)
        }

        val retVal = space.mkDOr(newArgs.build())


        return retVal

    }

    override fun _setIsSmooth() {
        //        assert smooth == null;
        //        assert isSmooth == null;
        _smooth = this
        _isSmooth = true
    }

    override val isSmooth: Boolean
        get() {
            if (_isSmooth == null) {
                _isSmooth = computeIsSmooth()
            }
            return _isSmooth!!
        }

    private fun computeIsSmooth(): Boolean {
        val parentVars = vars
        for (child in args) {
            val childVars = child.vars
            if (parentVars != childVars) {
                return false
            }
            if (!child.isSmooth) {
                return false
            }
        }
        return true
    }


    override fun computeValue(cube: Cube): Int {
        var totalValue = 0
        for (child in args) {
            val childValue = child.computeValue(cube)
            totalValue += childValue
        }
        return totalValue
    }

    override fun computeValue(): Int {
        assert(_space.pics != null)
        if (this._value == null) {
            var totalValue = 0
            for (child in args) {
                val childValue = child.computeValue()
                totalValue += childValue
            }
            this._value = totalValue
        }
        return this._value!!
    }

    override fun computeSat(lit: Lit): Boolean {
        for (arg in args) {
            val sat = arg.computeSat(lit)
            if (sat) return true
        }
        return false
    }

    override fun computeSat(cube: Cube): Boolean {
        for (arg in args) {
            val sat = arg.computeSat(cube)
            if (sat) return true
        }
        return false
    }

    override fun computeSat(trueVars: VarSet): Boolean {
        for (arg in args) {
            val sat = arg.computeSat(trueVars)
            if (sat) return true
        }
        return false
    }

    override fun computeSatCount1(lit: Lit): Long {
        var satCount: Long = 0

        val parentCareVars = vars

        for (child in args) {

            val satCountWithOutDcs = child.computeSatCount1(lit)

            val childCareVars = child.vars
            val dontCares = parentCareVars.minus(childCareVars)

            val dcCount = dontCares.size

            val dcSatCount = Exp.computeDcPermCount(dcCount).toLong()

            if (dcSatCount < 0) {
                throw IllegalStateException()
            }

            val satCountWithDcs = satCountWithOutDcs * dcSatCount

            satCount += satCountWithDcs


        }

        return satCount
    }

    override fun computeIsSat(): Boolean {
        for (arg in args) {
            val sat = arg.isSat
            if (sat) {
                return true
            }
        }
        return false
    }


    override fun copyToOtherSpace(destSpace: Space): Exp {
        return if (space === destSpace) {
            this
        } else Exp.copyArgsExpToOtherSpace(destSpace, op, argIt)
    }

    companion object {

        private val O_SP = "O "
    }

    //    @Override
    //    public Set<Lit> getLits() {
    //        ImmutableSet.Builder<Lit> b = ImmutableSet.builder();
    //        for (Exp arg : args) {
    //            b.addAll(arg.getLits());
    //        }
    //        return b.build();
    //    }

    //    @Override
    //    public boolean isLitMatched() {
    //        for (Exp arg : args) {
    //            if (!arg.isLitMatched()) {
    //                return false;
    //            }
    //        }
    //
    //        return true;
    //    }

}

